sympy/polys/polymatrix.py

```diff
@@ -1,87 +1,229 @@
     >>> v2.ring
-    ZZ
+    QQ[x]
     >>> pm2*v2
-    Matrix([[Poly(x**2, x, domain='QQ')]])
+    PolyMatrix([[x**2]], (x,))
 
     """
-    _class_priority = 10
-    # we don't want to sympify the elements of PolyMatrix
-    _sympify = staticmethod(lambda x: x)
-
-    def __init__(self, *args, ring=EX, **kwargs):
-        # kwargs are consumed by __new__, so need to be allowed here.
-        # if any non-Poly element is given as input then
-        # 'ring' defaults 'EX'
-        if all(isinstance(p, Poly) for p in self._mat) and self._mat:
-            domain = tuple([p.domain[p.gens] for p in self._mat])
-            ring = domain[0]
-            for i in range(1, len(domain)):
-                ring = ring.unify(domain[i])
-        self.ring = ring
-
-    def _eval_matrix_mul(self, other):
-        self_cols = self.cols
-        other_rows, other_cols = other.rows, other.cols
-        other_len = other_rows*other_cols
-        new_mat_rows = self.rows
-        new_mat_cols = other.cols
-
-        new_mat = [0]*new_mat_rows*new_mat_cols
-
-        if self.cols != 0 and other.rows != 0:
-            mat = self._mat
-            other_mat = other._mat
-            for i in range(len(new_mat)):
-                row, col = i // new_mat_cols, i % new_mat_cols
-                row_indices = range(self_cols*row, self_cols*(row+1))
-                col_indices = range(col, other_len, other_cols)
-                vec = (mat[a]*other_mat[b] for a,b in zip(row_indices, col_indices))
-                # 'Add' shouldn't be used here
-                new_mat[i] = sum(vec)
-
-        return self.__class__(new_mat_rows, new_mat_cols, new_mat, copy=False)
-
-    def _eval_scalar_mul(self, other):
-        mat = [Poly(a.as_expr()*other, *a.gens) if isinstance(a, Poly) else a*other for a in self._mat]
-        return self.__class__(self.rows, self.cols, mat, copy=False)
-
-    def _eval_scalar_rmul(self, other):
-        mat = [Poly(other*a.as_expr(), *a.gens) if isinstance(a, Poly) else other*a for a in self._mat]
-        return self.__class__(self.rows, self.cols, mat, copy=False)
 
+    def __new__(cls, *args):
+        if isinstance(args[0], list):
+            elements, args = args[0], args[1:]
+            if isinstance(elements[0], (list, tuple)):
+                rows, cols = len(elements), len(elements[0])
+                items = [e for row in elements for e in row]
+            else:
+                rows, cols = len(elements), 1
+                items = elements
+        elif [type(a) for a in args[:3]] == [int, int, list]:
+            rows, cols, items, args = args[0], args[1], args[2], args[3:]
+        elif [type(a) for a in args[:3]] == [int, int, type(lambda: 0)]:
+            rows, cols, func, args = args[0], args[1], args[2], args[3:]
+            items = [func(i, j) for i in range(rows) for j in range(cols)]
+        else:
+            1/0
+        return cls.from_list(rows, cols, items, *args)
```

Comment:
What if there are several args? Should this be just `args`?

Reply:
I guess if someone does
```
PolyMatrix(items, x, y)
```
then it would be `*args`. For Risch it seems that the Polys are always univariate.

I'm not really sure what sort of constructor options and syntax should be allowed. The main change here is the need to provide a generator if the inputs are not given as Poly.